### PR TITLE
core + editoast + front: simple implementation of the BP/CC

### LINKED ISSUE #4282: add to SimulationResult the BP/CC zones seen on the path (for the linear display part)
## Description :
The back should send infos for the linear diplay about the zones without electrification included in the path. In principle, the zone should contain the announce, the execution and the end/REV panels. The part of the zone displayed as neutral zone is in between the execution and the end/REV panel. For simplicity purposes and for the moment, we consider the zone starting at the announce. This will be corrected later.

## Implem:

- [x] Add the data in the same format as `ElectrificationConditions` but in another field. 
- [x] Update `core` and `editoast` openapi.yaml
- [x] Update `api` and `editoast` to return the new field of the result

### PR DESCRIPTION
**Pull request by Castavo**
In this PR:
- the simulation takes into account the BP/CC zones (coasting from the EXE panel to the FIN panel)
- display the results in the GEV

closes: #4282 
closes: #4365

### CODE REVIEW

**Comment by Castavo** [comment]
We will need to update editoast's openAPI, after #4303 has been merged

**Review by flomonster — APPROVED** [review]
The changes to small infra generator look good to me.

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPath.java
Comment (multun) [inline core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPath.java]: I would rather have dead sections be bundled with the catenary mode; my understanding is that we're not supposed to draw power within dead sections anyway, so it feels like to code will be simpler if we have both at hand at once

File: core/openapi.yaml
Comment (multun) [inline core/openapi.yaml:204]: shouldn't it be is_pantograph_drop?

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPath.java
Comment (multun) [inline core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPath.java]: My understanding is that for now, dead sections are only passed through, and don't change the actual simulation results. Is that correct?

File: core/openapi.yaml
Comment (Castavo) [reply core/openapi.yaml:204]: Fair question, I had the idea more of a "this is a zone that intimates the order to drop the pantograph" rather than "a zone where pantograph is dropped".
With my explanation, do you still think that `is_pantograph_drop` is more sensible ?

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPath.java
Comment (Castavo) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPath.java]: That is correct
The simulation results should be changed in the upcoming weeks.
This PR will allow the front devs to works on the linear display.

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPath.java
Comment (Castavo) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPath.java]: Fine idea !

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPath.java
Comment (clarani) [inline core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPath.java]: Comment to adapt ?

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_utils/RangeMapUtils.java
Comment (clarani) [inline core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_utils/RangeMapUtils.java]: Can't do that, you won't merge the 2 last ranges even if they have the same value

File: core/envelope-sim/src/test/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPathTest.java
Comment (clarani) [inline core/envelope-sim/src/test/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPathTest.java]: Can you explain this test ?

File: core/envelope-sim/src/test/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPathTest.java
Comment (clarani) [inline core/envelope-sim/src/test/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPathTest.java]: Needed ?

File: core/src/main/java/fr/sncf/osrd/train/RollingStock.java
Comment (clarani) [inline core/src/main/java/fr/sncf/osrd/train/RollingStock.java]: *electrification conditions that matched

File: core/src/main/java/fr/sncf/osrd/standalone_sim/result/ElectrificationConditionsRange.java
Comment (clarani) [inline core/src/main/java/fr/sncf/osrd/standalone_sim/result/ElectrificationConditionsRange.java]: Why final or not final ?

Comment (Castavo) [reply core/src/main/java/fr/sncf/osrd/standalone_sim/result/ElectrificationConditionsRange.java]: I made it not final so that I could update the stop if the next range computed in fact holds the same value

File: core/envelope-sim/src/test/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPathTest.java
Comment (Castavo) [reply core/envelope-sim/src/test/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPathTest.java]: I thought you asked if the change was needed at first, your message isn't very explicit 
Indeed the variable declaration isn't needed :grin:

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_utils/RangeMapUtils.java
Comment (Castavo) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_utils/RangeMapUtils.java]: I really think the algorithm works, I'll write a test to enforce it

File: core/envelope-sim/src/test/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPathTest.java
Comment (Castavo) [reply core/envelope-sim/src/test/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPathTest.java:44]: I'll add a comment

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPath.java
Comment (Castavo) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPath.java]: Like this ?
```suggestion
     * Get the electrification related data for a given power class and power restriction map.
```

File: core/src/main/java/fr/sncf/osrd/standalone_sim/result/ElectrificationConditionsRange.java
Comment (clarani) [reply core/src/main/java/fr/sncf/osrd/standalone_sim/result/ElectrificationConditionsRange.java]: Thanks for the explanation !

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/TrainPhysicsIntegrator.java
Comment (axrolld) [inline core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/TrainPhysicsIntegrator.java]: Isn't there a way we make sure `position` is in an acceptable range before this line ?

File: core/openapi.yaml
Comment (axrolld) [reply core/openapi.yaml:204]: What about `requires_pantograph_drop` ?

**Review by axrolld — CHANGES_REQUESTED** [review]

File: core/openapi.yaml
Comment (Castavo) [reply core/openapi.yaml:204]: Yes, I propose we change the names in one large refactoring haul later

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_utils/RangeMapUtils.java
Comment (clarani) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_utils/RangeMapUtils.java]: My bad, I think the algo is working ! A test would be super nice :)

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/TrainPhysicsIntegrator.java
Comment (Castavo) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/TrainPhysicsIntegrator.java]: Do you mean to ask if there is a way to avoid having to clamp the value through the algorithm itself, or do you mean clamping the value earlier in the function ?

Comment (axrolld) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/TrainPhysicsIntegrator.java]: I mean at this point i really do not expect to have a positions anywhere but between the beginning and the end of the track. I'm surprised the clamp is necessary.

Comment (Castavo) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/TrainPhysicsIntegrator.java]: Well I was surprised too you see
It's necessary because of the positionDelta which leads to ever so slightly exit the path.
That's why it's applied in `getWeightForce` for example

Comment (axrolld) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/TrainPhysicsIntegrator.java]: But how come the positionDelta leads to an exit of the path now ? I didn't have a look in depth in the rest of the PR but if this change is necessary that _might_ need something's wrong somewhere (i'm really not saying that's the case, i'm just wondering).

I agree with the change anyways, as it cannot lead to wrong results, but i'm still surprised

**Review by axrolld — APPROVED** [review]

File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/TrainPhysicsIntegrator.java
Comment (Castavo) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/TrainPhysicsIntegrator.java]: In fact it always did before, it's just that before the mapping extended out of the path, and now not anymore

Comment (axrolld) [reply core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/TrainPhysicsIntegrator.java]: Alright, that's clearer, thanks for investigating ! :)

File: front/src/reducers/osrdsimulation/types.ts
Comment (anisometropie) [inline front/src/reducers/osrdsimulation/types.ts]: use the `osrdEditoastApi.ts` definitions for `Electrified ` `Neutral` `NonElectrified`

File: front/src/reducers/osrdsimulation/types.ts
Comment (anisometropie) [inline front/src/reducers/osrdsimulation/types.ts]: for the TS types we never use `Object`, `Number`, `String`, `Symbol`, `Array` but the lowercase counterpart
```suggestion
  type: object;
```

File: front/src/applications/operationalStudies/components/ManageTrainSchedule/SubmitConfAddTrainSchedule.tsx
Comment (nicolaswurtz) [inline front/src/applications/operationalStudies/components/ManageTrainSchedule/SubmitConfAddTrainSchedule.tsx]: Huuuu ?

Comment (anisometropie) [inline front/src/applications/operationalStudies/components/ManageTrainSchedule/SubmitConfAddTrainSchedule.tsx]: remove

File: front/src/reducers/osrdconf/index.ts
Comment (nicolaswurtz) [inline front/src/reducers/osrdconf/index.ts]: Why not using `powerRestrictionRanges` too for `action` ?

File: front/src/reducers/osrdconf/selectors.ts
Comment (nicolaswurtz) [inline front/src/reducers/osrdconf/selectors.ts]: Why not using `getPowerRestrictionRanges` ?

**Review by nicolaswurtz — CHANGES_REQUESTED** [review]

File: editoast/openapi.yaml
Comment (anisometropie) [inline editoast/openapi.yaml]: discriminator `object_type` must be on all the elements of the union  `Electrified ` `Neutral` `NonElectrified`
```suggestion
    Electrified:
      properties:
        mode:
          type: string
        mode_handled:
          type: boolean
        profile:
          type: string
          nullable: true
        profile_handled:
          type: boolean
        object_type:
          type: string
          enum: ["electrified"]
      required: [mode, mode_handled, profile_handled]
```

**Review by anisometropie — CHANGES_REQUESTED** [review]

File: front/src/applications/operationalStudies/components/ManageTrainSchedule/SubmitConfAddTrainSchedule.tsx
Comment (Math-R) [reply front/src/applications/operationalStudies/components/ManageTrainSchedule/SubmitConfAddTrainSchedule.tsx]: Done

File: front/src/reducers/osrdconf/index.ts
Comment (Math-R) [reply front/src/reducers/osrdconf/index.ts]: Done

File: front/src/reducers/osrdconf/selectors.ts
Comment (Math-R) [reply front/src/reducers/osrdconf/selectors.ts]: Done

File: front/src/reducers/osrdsimulation/types.ts
Comment (Math-R) [reply front/src/reducers/osrdsimulation/types.ts]: Done !

File: front/src/reducers/osrdsimulation/types.ts
Comment (Math-R) [reply front/src/reducers/osrdsimulation/types.ts]: Done

File: front/src/applications/operationalStudies/components/ManageTrainSchedule/SubmitConfAddTrainSchedule.tsx
Comment (Math-R) [reply front/src/applications/operationalStudies/components/ManageTrainSchedule/SubmitConfAddTrainSchedule.tsx]: done

File: editoast/openapi.yaml
Comment (Math-R) [reply editoast/openapi.yaml]: done

**Review by anisometropie — APPROVED** [review]

**Review by nicolaswurtz — APPROVED** [review]

**Review by multun — APPROVED** [review]

**Pull request merged**
